src/cpu/Dasm.cc

```diff
@@ -111,8 +111,38 @@ unsigned dasm(const MSXCPUInterface& interface, uint16_t pc, std::span<uint8_t,
 	return i;
 }
 
-unsigned instructionLength(const MSXCPUInterface& interface, uint16_t pc,
-                           EmuTime::param time)
+static std::span<uint8_t> fetchInstruction(const MSXCPUInterface& interface, uint16_t addr,
+                                           std::span<uint8_t, 4> buffer, EmuTime::param time) {
+	// read 1st byte into buffer
+	buffer[0] = interface.peekMem(addr, time);
+	// calculate the instruction length (first part)
+	auto len1 = instr_len_tab[buffer[0]];
+	if (len1 < 4) {
+		// if regular instruction, read 0 to 2 more bytes and finish (3 bytes total)
+		for (unsigned i = 1; i < len1; ++i) {
+			buffer[i] = interface.peekMem(addr + narrow_cast<uint16_t>(i), time);
+		}
+		return buffer.subspan(0, len1);
+	}
+
+	// if DD, ED or FD instruction, read 0 to 3 more bytes into buffer (4 bytes total)
+	buffer[1] = interface.peekMem(addr + 1, time);
+	auto len2 = instr_len_tab[64 * len1 + buffer[1]];
+	for (unsigned i = 2; i < len2; ++i) {
+		buffer[i] = interface.peekMem(addr + narrow_cast<uint16_t>(i), time);
+	}
+	return buffer.subspan(0, len2);  // return the given buffer, but trimmed to the correct length
+}
+
+unsigned dasm(const MSXCPUInterface& interface, uint16_t pc, std::span<uint8_t, 4> buf,
+              std::string& dest, EmuTime::param time,
+              function_ref<void(std::string&, uint16_t)> appendAddr)
+{
+	return dasm(fetchInstruction(interface, pc, buf, time), pc, dest, appendAddr);
+}
```

Comment:
If you rework `dasm()` to return `void`. This can become:
```
auto opcodes = fetchInstruction(...)
dasm(opcodes, ...);
return opcodes.size();
```

Or alternatively you can change the return type from `unsigned` to `std::span<uint8_t>`. So similar to how the return is handled in `fetchInstruction()`.